### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by arinaldi118**
## Summary

- Reimplemented tokens with expiration time.
- Users with multiple sessions.
- Implemented new model for sessions.
- Implemented user logout.

## Trello Cards
[https://trello.com/c/yihMbWlo/5-autenticaci%C3%B3n](url)

### CODE REVIEW

File: app/controllers/sessions.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    : {};
+
+  return sessionService
+    .create(session)
+    .then(u => {
+      res.status(201);
+      res.end();
+    })
+    .catch(err => {
+      next(errors.defaultError(err));
```
Comment (igna92ts) [inline app/controllers/sessions.js]: This is completely optional but it looks nicer if in a `catch` inside the service you throw a defaultError so the here you can do 
```
.catch(next);
```

File: app/controllers/sessions.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    });
+};
+
+exports.logout = (req, res, next) => {
+  return sessionService
+    .delete(req.body.email, req.headers.authorization)
```
Comment (igna92ts) [inline app/controllers/sessions.js:25]: where are you validating that both attributes are present

File: app/controllers/sessions.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      res.end();
+    })
+    .catch(err => {
+      next(errors.defaultError(err));
+    });
+};
+
+exports.logoutAll = (req, res, next) => {
+  return sessionService
+    .deleteAll(req.body.email, req.headers.authorization)
```
Comment (igna92ts) [inline app/controllers/sessions.js:35]: same

File: app/middlewares/auth.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+const sessionManager = require('./../services/sessionManager');
+const orm = require('./../orm');
+const errors = require('./../errors');
+const sessionService = require('./../services/sessions');
+
+exports.secure = (req, res, next) => {
+  const auth = req.headers[sessionManager.HEADER_NAME];
```
Comment (igna92ts) [inline app/middlewares/auth.js:7]: same

File: app/services/sessionManager.js [hunk 5]
```diff
@@ -1,14 +1,16 @@
 const jwt = require('jwt-simple'),
   config = require('./../../config');
+const errors = require('./../errors');
```
Comment (igna92ts) [inline app/services/sessionManager.js]: include this with the same const declaration like you did with `config`

File: test/users.js [hunk 6]
```diff
@@ -223,4 +245,103 @@ describe('users test', () => {
...
+          }, 1000);
+        });
+      });
+    });
+  });
+
+  describe('/users/logout/all POST', () => {
+    it('Should fail because of invalid token', done => {
+      return successfulLogin().then(res => {
+        res.should.have.status(201);
```
Comment (igna92ts) [inline test/users.js:339]: is the second login necessary here?

File: test/users.js [hunk 7]
```diff
@@ -223,4 +245,103 @@ describe('users test', () => {
...
+            });
+          });
+        }, 1000);
+      });
+    });
+    it('Should successful because sessions database is empty', done => {
+      return successfulLogin().then(res => {
+        res.should.have.status(201);
+        setTimeout(() => {
+          return successfulLogin().then(res2 => {
```
Comment (igna92ts) [inline test/users.js:389]: same

**Review by igna92ts — CHANGES_REQUESTED** [review]

File: test/users.js (hunk 6 above)
Comment (arinaldi118) [reply test/users.js:339]: Yes. I want to see the behavior with multiple sessions.

File: test/users.js (hunk 7 above)
Comment (arinaldi118) [reply test/users.js:389]: Same

File: test/users.js (hunk 6 above)
Comment (igna92ts) [reply test/users.js:339]: but you included this in all the subsequent tests, didnt you check this behaviour specifically in a previous one??

**Review by igna92ts — APPROVED** [review]

**Pull request merged**